### PR TITLE
[Feature] Support structural tag templates.

### PR DESCRIPTION
**Pull request by Seven-Streams**
This PR supports the structural tag templates, provides a method to generate a grammar from a structural tag template and the given values, and provides built-in structural tag templates for several models' tool-calling formats.

### CODE REVIEW

File: python/xgrammar/grammar.py [hunk 1]
```diff
@@ -431,3 +432,65 @@ def deserialize_json(json_string: str) -> "Grammar":
             When the __VERSION__ field in the JSON string is not the same as the current version.
         """
         return Grammar._create_from_handle(_core.Grammar.deserialize_json(json_string))
+
+    @staticmethod
+    def apply_structural_tag_template(
+        template_json_str: Union[str, Dict[str, Any]], **kwargs: List[Dict[str, Any]]
```
Comment (Ubospica) [inline python/xgrammar/grammar.py]:
```suggestion
        template_json_str: Union[str, Dict[str, Any], StructuralTag], **kwargs: List[Dict[str, Any]]
```

File: python/xgrammar/structural_tag.py [hunk 2]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
     "StructuralTagItem",
     "StructuralTag",
 ]
+
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]: StructuralTag.apply_structural_tag_template(...) -> StructuralTag

builtin_structural_tag_template.py

File: python/xgrammar/structural_tag.py [hunk 3]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
...
+    """Register a structural tag format."""
+
+    def decorator(func):
+        _structural_tag_registry[name] = func
+        return func
+
+    return decorator
+
+
+def get_builtin_template_structural_tag(format_type: str) -> Dict[str, Any]:
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
def get_builtin_structural_tag_template(format_name: str) -> StructuralTag:
```

File: python/xgrammar/structural_tag.py [hunk 4]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
...
+    """
+    func = _structural_tag_registry.get(format_type)
+    if func is None:
+        support_types = list(_structural_tag_registry.keys())
+        raise ValueError(f"Unknown format type: {format_type}, support types: {support_types}")
+    return func()
+
+
+@_register_template_structural_tag_format("Llama")
+def get_llama_style_template_structural_tag() -> Dict[str, Any]:
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
def _get_llama_structural_tag_template() -> Dict[str, Any]:
```

File: python/xgrammar/structural_tag.py [hunk 5]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
...
+
+    """
+    func = _structural_tag_registry.get(format_type)
+    if func is None:
+        support_types = list(_structural_tag_registry.keys())
+        raise ValueError(f"Unknown format type: {format_type}, support types: {support_types}")
+    return func()
+
+
+@_register_template_structural_tag_format("Llama")
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
@_register_template_structural_tag_format("llama")
```

File: python/xgrammar/structural_tag.py [hunk 6]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
...
+                    "begin": "<｜tool▁calls▁begin｜><｜tool▁call▁begin｜>{tools[].name}<｜tool▁sep｜>",
+                    "content": {"type": "json_schema", "json_schema": "{tools[].parameters}"},
+                    "end": "<｜tool▁call▁end｜>",
+                }
+            ],
+        },
+    }
+
+
+@_register_template_structural_tag_format("Qwen_Coder")
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
@_register_template_structural_tag_format("qwen_coder")
```

File: python/xgrammar/structural_tag.py [hunk 7]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
     "StructuralTagItem",
     "StructuralTag",
 ]
+
+# ---------- Template Structural Tag ----------
+
+_structural_tag_registry = {}
+
+
+def _register_template_structural_tag_format(name: str):
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
def _register_structural_tag_template(name: str):
```

File: python/xgrammar/structural_tag.py (hunk 3 above)
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]:
```suggestion
def get_builtin_structural_tag_template(format_type: str) -> Dict[str, Any]:
```

File: docs/tutorials/template_structural_tag.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
+# Template Structural Tags
+
+Based on the [structural tags](./structural_tag.md), template structural tags provides a more convineient method for users to generate grammars to describe the grammar to constrain the LLMs' output.
+In general, Template structural tags support placeholders in the structural tags, which are in the form of `{lists[].arg}`. These values will be automatically expanded with the user's input values. Users can use `xgrammar.Grammar.apply_structural_tag_template(template_json_str: Union[str, Dict[str, Any]], **kwargs: List[Dict[str, Any]])` to generate a `xgrammar.Grammar`. Here, `template_json_str` is the template structural tags, and kwargs are a series of values. For example, the template structural tag `stag` contains `{A[].name}` and `{A[].age}`, then the users can use `xgrammar.Grammar.apply_structural_tag_template(stag, A=A)` to generate the grammar, where `A=[{"name":..., "age":...},.{"name":...,"age":...}, ...]`. This function will replace the placeholders automatically.
+
```
Comment (Ubospica) [inline docs/tutorials/template_structural_tag.md]: 1. Static, for model designer, general (reasoning, tool calling, etc.)

Provide a code snippet for usage
1. llama reasoning + tool calling. Declare a stag object; apply to StructuralTag; Compile to grammar

2. Usage: 1) huggingface logits adapter; 2) SGLang/vLLM inference API (initialize the engine; send request)


Template Semantics:
1. Substitution: const string (part), json schema, full  "{begin} a dog{end}". In string, for original {, use {{. For regex/grammar, only full substitution. {{ }}
\{
regex: "abc{1,2}"
regex: "{{}}" # -> {}; "
3. Dict indices: "{begin.begin} a dog{begin.end}"
4. List indices: "{begin.begin} a dog{begin.end}". LCA semantics.

File: python/xgrammar/structural_tag.py [hunk 9]
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
...
+def get_qwen_style_template_structural_tag() -> Dict:
+    """Get Qwen style structural tag format.
+
+    Returns
+    -------
+    Dict[str, Any]
+        A template structural tag format dictionary.
+        This format is used by Qwen3 and other models that follow the same style.
+    """
+    return {
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py]: add reasoning format to it

1. qwen
2. qwen_non_thinking

File: python/xgrammar/grammar.py [hunk 10]
```diff
@@ -431,3 +432,65 @@ def deserialize_json(json_string: str) -> "Grammar":
             When the __VERSION__ field in the JSON string is not the same as the current version.
         """
         return Grammar._create_from_handle(_core.Grammar.deserialize_json(json_string))
+
+    @staticmethod
+    def apply_structural_tag_template(
```
Comment (Ubospica) [inline python/xgrammar/grammar.py]: Grammar.from_structural_tag_template(...)

File: docs/tutorials/template_structural_tag.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    "value": "{A[].begin} a dog{A[].end}"
+}
+```
+
+Is allowed. And the output is constrained to `It is a dog.` and `Is it a dog?`. However, if the provided values are `Begin=[{"word": "It is"}, {"word": "Is it"}], End=[{"word": "."}, {"word": "?"}]`, and the template is
+
+```json
+{
+    "type": "const_string",
+    "value": "{Begin[].word} a dog{End[].word}"
```
Comment (Ubospica) [inline docs/tutorials/template_structural_tag.md]:
```suggestion
    "value": "{value.begin1.begin2} a dog{value.begin1.end2}"
```

apply(begin={"begin1": {"begin2": ..., "end2": ...})

File: cpp/structural_tag.h [hunk 12]
```diff
@@ -55,35 +56,41 @@ struct ConstStringFormat {
...
   std::string xml_schema;
   QwenXmlParameterFormat(std::string xml_schema) : xml_schema(std::move(xml_schema)) {}
+  std::string ToString() const { return "{type: qwen_xml, xml_schema: " + xml_schema + "}"; }
 };
 
 struct GrammarFormat {
   static constexpr const char* type = "grammar";
   std::string grammar;
   GrammarFormat(std::string grammar) : grammar(std::move(grammar)) {}
+  std::string ToString() const { return "{type: grammar, grammar: " + grammar + "}"; }
```
Comment (Ubospica) [inline cpp/structural_tag.h]: ToJSON

File: cpp/structural_tag.cc [hunk 13]
```diff
@@ -23,6 +33,26 @@ namespace xgrammar {
 // Short alias for the error type.
 using ISTError = InvalidStructuralTagError;
 
+// Format: {{var_name([])? (.var_name([])?)*}}
+const std::regex full_placeholder_regex =
```
Comment (Ubospica) [inline cpp/structural_tag.cc:36]: Time can be further checked. Could better be local / class local

File: python/xgrammar/builtin_structural_tag_template.py [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+    TagFormat,
+    TriggeredTagsFormat,
+)
+
+# ---------- Template Structural Tag ----------
+
+_structural_tag_registry = {}
+
+
+def _register_template_structural_tag_format(name: str):
```
Comment (Ubospica) [inline python/xgrammar/builtin_structural_tag_template.py]:
```suggestion
def _register_structural_tag_template(name: str):
```

File: python/xgrammar/builtin_structural_tag_template.py [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+    """Register a structural tag format."""
+
+    def decorator(func):
+        _structural_tag_registry[name] = func
+        return func
+
+    return decorator
+
+
+def get_builtin_template_structural_tag(format_type: str) -> StructuralTag:
```
Comment (Ubospica) [inline python/xgrammar/builtin_structural_tag_template.py]:
```suggestion

SupportedTemplateName = Literal["llama", "qwen", ...]

def get_builtin_structural_tag_template(format_type: Union[SupportedTemplateName, str]) -> StructuralTag:
```

File: python/xgrammar/structural_tag.py [hunk 16]
```diff
@@ -29,7 +29,7 @@ class JSONSchemaFormat(BaseModel):
 
     type: Literal["json_schema"] = "json_schema"
     """The type of the format."""
-    json_schema: Union[bool, Dict[str, Any]]
+    json_schema: Union[bool, Dict[str, Any], str]
     """The JSON schema."""
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py:35]:
```suggestion
    """The JSON schema. In the structural tag template, this is a string as the placeholder name."""
```

File: python/xgrammar/structural_tag.py [hunk 17]
```diff
@@ -61,7 +61,7 @@ class QwenXMLParameterFormat(BaseModel):
     type: Literal["qwen_xml_parameter"] = "qwen_xml_parameter"
     """The type of the format."""
 
-    json_schema: Union[bool, Dict[str, Any]]
+    json_schema: Union[bool, Dict[str, Any], str]
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py:66]: ditto

File: docs/tutorials/structural_tag_template.md [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+
+4. `triggered_tags`, `tags_with_separator`
+
+Template placeholders are not allowed in `triggers` and `separators`.
+
+5. `sequence`, `or`, `any_text`
+
+Template placeholders cannot be directly contained in these formats.
+
+## InValid Structural Tag Templates
```
Comment (Ubospica) [inline docs/tutorials/structural_tag_template.md]:
```suggestion
## Common Invalid Patterns
```

File: python/xgrammar/structural_tag.py [hunk 19]
```diff
@@ -61,7 +61,7 @@ class QwenXMLParameterFormat(BaseModel):
...
     """The JSON schema for the parameters of the function calling."""
```
Comment (Ubospica) [inline python/xgrammar/structural_tag.py:68]: StructuralTag.from_template(template, **kwargs)

File: docs/tutorials/structural_tag_template.md [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+
+Each `const_string` format can contain multiple placeholders, but they must be from the same value mapping. for example, `A=[{"begin": "It is", "end":"."}, {"begin": "Is it", "end": "?"}]`, and the template is:
+
+```json
+{
+    "type": "const_string",
+    "value": "{{A[].begin}} a dog{{A[].end}}"
+}
+```
+
```
Comment (Ubospica) [inline docs/tutorials/structural_tag_template.md:21]: Python code: one line, 
```python
structural_tag = StructuralTag.from_template(..., A=xxx)
```

Converted Structural Tag:
```json
{
  "type": "or",
   ...
}
```

File: docs/tutorials/structural_tag_template.md [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+### Type 1: Mingled Scope
+
+Not all the structural tag templates are valid. For example, the mingled formats mentioned above are not valid structural tag templates. Besides, there are some other situations where the template structural tag cannot be compiled. For example:
+
+```json
+{
+    "type": "sequence",
+    "elements": [
+            {
```
Comment (Ubospica) [inline docs/tutorials/structural_tag_template.md:249]: indentation

File: docs/api/python/structural_tag.rst [hunk 22]
```diff
@@ -72,3 +72,8 @@ Combinatorial Formats
 .. autopydantic_model:: TagsWithSeparatorFormat
    :show-inheritance:
    :exclude-members: model_config
+
+Template Structural Tag
```
Comment (Ubospica) [inline docs/api/python/structural_tag.rst]:
```suggestion
Structural Tag Template
```

**Review by Ubospica — COMMENTED** [review]
Thanks! Generally looks good. Let's measure conversion time & compile time for 5 tools & 20 tools

File: cpp/json_schema_converter.cc [hunk 23]
```diff
@@ -979,7 +981,7 @@ std::string JSONSchemaConverter::VisitConst(
 }
 
 std::string JSONSchemaConverter::VisitEnum(
-    const picojson::object& schema, const std::string& rule_name
+    const picojson::object& schema, const std::string& rule_name, const JSONFormat json_format
```
Comment (Ubospica) [inline cpp/json_schema_converter.cc]:
```suggestion
    const picojson::object& schema, const std::string& rule_name, JSONFormat json_format
```

**Review by Ubospica — COMMENTED** [review]
Thanks. Let's update tests and we can merge

File: cpp/structural_tag.cc [hunk 24]
```diff
@@ -24,6 +33,26 @@ namespace xgrammar {
...
+    std::regex(R"(\{\{([a-zA-Z_][a-zA-Z0-9_]*)(\[\])?(\.([a-zA-Z_][a-zA-Z0-9_]*)(\[\])?)*\}\})");
+const std::regex placeholder_regex = std::regex(R"(([a-zA-Z_][a-zA-Z0-9_]*)(\[\])?)");
+
+std::string FormatToJSONStr(const Format& format) {
+  return std::visit([&](auto&& arg) -> std::string { return arg.ToJSON(); }, format);
+}
+
```
Comment (Ubospica) [inline cpp/structural_tag.cc:43]:
```suggestion
picojson::value FormatToJSONStr(const Format& format) {
  return std::visit([&](auto&& arg) -> picojson::value { return arg.ToJSON(); }, format);
}
```

File: efficiency_test.py [hunk 25]
```diff
@@ -0,0 +1,149 @@
+import time
```
Comment (Ubospica) [inline efficiency_test.py]: just remove it

File: docs/tutorials/structural_tag_template.md [hunk 26]
```diff
@@ -0,0 +1,382 @@
+# Structural Tag Templates
+
+Based on the [structural tags](./structural_tag.md), structural tag templates provides a more convineient method for users to generate grammars to describe the grammar to constrain the LLMs' output.
+In general, structural tag templates support placeholders in the structural tags, which are in the form of `\{\{(var_name([])?)(\.var_name([])?)\}\}`, such as `{{A.name}}`, `{{A[].name}}`, `{{A.name[].other}}`. `var_name[]` means that the value is an array, and it will be automatically expanded with the users' input. `var_name` means that the values is a terminal value, and it will be automatically replaced with the input. Users can use `xgrammar.Grammar.from_structural_tag_template()` to generate a `xgrammar.Grammar`. Here, `template_json_str` is the structural tag templates, and kwargs are a series of values. For example, the structural tag template `stag` contains `{{A[].name}}` and `{{A[].age}}`, then the users can use `xgrammar.Grammar.from_structural_tag_template(stag, A=A)` to generate the grammar, where `A=[{"name":..., "age":...},.{"name":...,"age":...}, ...]`. This function will replace the placeholders automatically.
```
Comment (Ubospica) [inline docs/tutorials/structural_tag_template.md:7]:
```suggestion

In general, structural tag templates support placeholders in the structural tags, which are in the form of `\{\{(var_name([])?)(\.var_name([])?)\}\}`, such as `{{A.name}}`, `{{A[].name}}`, `{{A.name[].other}}`. `var_name[]` means that the value is an array, and it will be automatically expanded with the users' input. `var_name` means that the values is a terminal value, and it will be automatically replaced with the input. Users can use `xgrammar.Grammar.from_structural_tag_template()` to generate a `xgrammar.Grammar`. Here, `template_json_str` is the structural tag templates, and kwargs are a series of values. For example, the structural tag template `stag` contains `{{A[].name}}` and `{{A[].age}}`, then the users can use `xgrammar.Grammar.from_structural_tag_template(stag, A=A)` to generate the grammar, where `A=[{"name":..., "age":...},.{"name":...,"age":...}, ...]`. This function will replace the placeholders automatically.
```

File: docs/tutorials/structural_tag_template.md [hunk 27]
```diff
@@ -0,0 +1,382 @@
+# Structural Tag Templates
+
+Based on the [structural tags](./structural_tag.md), structural tag templates provides a more convineient method for users to generate grammars to describe the grammar to constrain the LLMs' output.
```
Comment (Ubospica) [inline docs/tutorials/structural_tag_template.md]:
```suggestion
Structural Tags provide a way to describe complex output formats for LLMs. However, for agent applications, LLM's output format depends on dynamic runtime values, such as tool calling specifications, that can be diverse for every request. 

The Structural Tag Template provides an **unified and static** way to represent such dynamic formats: Users only need to provide one structural tag template, and simply use runtime arguments to specialize it to get dynamic formats.

```
